consumer-services/src/main/scala/za/co/absa/spline/consumer/service/internal/AttributeDependencySolver.scala

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2020 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.spline.consumer.service.internal
+
+import java.util.UUID
+
+import za.co.absa.spline.consumer.service.internal.model.OperationWithSchema
+import za.co.absa.spline.consumer.service.model.AttributeDependencies
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+object AttributeDependencySolver {
+
+  def resolveDependencies(operations: Seq[OperationWithSchema], attributeID: UUID): AttributeDependencies = {
+
+    val operationsMap = operations.map(op => op._id -> op).toMap
+    val inputSchemaResolver = createInputSchemaResolver(operations)
+
+    def findDependenciesStart(
+      operation: OperationWithSchema,
+      attributeID: UUID
+    ): (Set[UUID], Set[String]) = {
+
+      if (operation.schema.contains(attributeID))
+        findTransitiveDependencies(operation, Set(attributeID))
+      else if (operation.childIds.isEmpty)
+        (Set.empty, Set.empty)
+      else
+        operation
+          .childIds
+          .map(operationsMap)
+          .map(findDependenciesStart(_, attributeID))
+          .reduce(elementwiseUnion(_, _))
+    }
+
+    def findTransitiveDependencies(
+      operation: OperationWithSchema,
+      attributeDependencies: Set[UUID]
+    ): (Set[UUID], Set[String]) = {
+
+      val dependencyMap = resolveDependencies(operation, inputSchemaResolver)
+      val updatedAttributeDependencies = attributeDependencies
+        .flatMap(dependencyMap.get(_))
+        .flatten
+        .union(attributeDependencies)
+
+      val (attrDep, opDep) = operation
+        .childIds
+        .map(operationsMap)
+        .filter(_.schema.exists(updatedAttributeDependencies))
+        .map(findTransitiveDependencies(_, updatedAttributeDependencies))
+        .reduceOption(elementwiseUnion(_, _))
+        .getOrElse((updatedAttributeDependencies, Set.empty[String]))
+
+      (attrDep, opDep + operation._id)
+    }
+
+    def elementwiseUnion[A, B](x: (Set[A], Set[B]), y: (Set[A], Set[B])): (Set[A], Set[B]) =
+      (x._1 union y._1, x._2 union y._2)
+
+
+    val writeOp = operations.find(_.extra("name") == "Write").get
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Usage of get on optional type.](https://app.codacy.com/manual/ABSA_OSS/spline/pullRequest?prid=4773990)